### ISSUE TITLE
Compiling OpenBLAS on Android with debug symbols

### ISSUE THREAD

**Issue by dugsmith** [xianyi/OpenBLAS#2692]
Hello,

I'm using NDK r20b to build OpenBLAS v0.3.4. I'd like to have debug symbols output in the compiled files so that I can see stack traces when there are crashes.

I'm seeing a behavior in `clang` that I don't understand. The `clang` statement generated by make does have the `-g` parameter needed to generate debug symbols. However, the subsequently executed `clang` statement has the `-g` stripped from it, and no debug symbols are in the compiled file. 

For example, here is one of the make-generated `clang` statements:

```sh
/Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/armv7a-linux-androideabi23-clang -g -v -marm -mfpu=vfp -mfloat-abi=softfp  -O2 -DMAX_STACK_ALLOC=2048 -marm -mfloat-abi=softfp -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.4\" -mfpu=neon  -march=armv7-a -DASMNAME=cblas_isamax -DASMFNAME=cblas_isamax_ -DNAME=cblas_isamax_ -DCNAME=cblas_isamax -DCHAR_NAME=\"cblas_isamax_\" -DCHAR_CNAME=\"cblas_isamax\" -DNO_AFFINITY -I.. -I. -UDOUBLE  -UCOMPLEX -DCBLAS -c -DUSE_ABS -UUSE_MIN imax.c -o cblas_isamax.o
```

And here is what is generated by the verbose output of `clang`:

```sh
Android (5220042 based on r346389c) clang version 8.0.7 (https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf643d27dbca1bb228957008617) (https://android.googlesource.com/toolchain/llvm 3c393fe7a7e13b0fba4ac75a01aa683d7a5b11cd) (based on LLVM 8.0.7svn)
Target: armv7a-unknown-linux-android23
Thread model: posix
InstalledDir: /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin
Found candidate GCC installation: /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x
Selected GCC installation: /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x
Candidate multilib: thumb;@mthumb
Candidate multilib: armv7-a;@march=armv7-a
Candidate multilib: armv7-a/thumb;@march=armv7-a@mthumb
Candidate multilib: .;
Selected multilib: armv7-a;@march=armv7-a
 "/Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang" -cc1 -triple armv7-unknown-linux-android23 -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name imax.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -masm-verbose -mconstructor-aliases -fuse-init-array -target-cpu generic -target-feature +soft-float-abi -target-feature -fp-only-sp -target-feature -d16 -target-feature +vfp3 -target-feature -fp16 -target-feature -vfp4 -target-feature -fp-armv8 -target-feature +neon -target-feature -crypto -target-abi aapcs-linux -mfloat-abi soft -fallow-half-arguments-and-returns -dwarf-column-info -debug-info-kind=limited -dwarf-version=4 -debugger-tuning=gdb -target-linker-version 241.9 -v -coverage-notes-file /Users/doug.smith/dev/prebuilt/OpenBLAS/kernel/arm/cblas_isamax.gcno -resource-dir /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/lib64/clang/8.0.7 -D MAX_STACK_ALLOC=2048 -D F_INTERFACE_GFORT -D NO_LAPACK -D NO_LAPACKE -D SMP_SERVER -D NO_WARMUP -D MAX_CPU_NUMBER=12 -D MAX_PARALLEL_NUMBER=1 -D "VERSION=\"0.3.4\"" -D ASMNAME=cblas_isamax -D ASMFNAME=cblas_isamax_ -D NAME=cblas_isamax_ -D CNAME=cblas_isamax -D "CHAR_NAME=\"cblas_isamax_\"" -D "CHAR_CNAME=\"cblas_isamax\"" -D NO_AFFINITY -I .. -I . -U DOUBLE -U COMPLEX -D CBLAS -D USE_ABS -U USE_MIN -internal-isystem /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../sysroot/usr/local/include -internal-isystem /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/lib64/clang/8.0.7/include -internal-externc-isystem /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../sysroot/usr/include/arm-linux-androideabi -internal-externc-isystem /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../sysroot/include -internal-externc-isystem /Users/doug.smith/Library/Android/sdk/ndk/20.1.5948944/toolchains/llvm/prebuilt/darwin-x86_64/bin/../sysroot/usr/include -O2 -Wall -fdebug-compilation-dir /Users/doug.smith/dev/prebuilt/OpenBLAS/kernel/arm -ferror-limit 19 -fmessage-length 212 -fno-signed-char -fobjc-runtime=gcc -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o cblas_isamax.o -x c imax.c
```

You can see that among other things, the `-cc1` attribute is added, the target architecture is added as a `-triple`, and many other attributes are added. I'm thinking they are added by the Android toolchain somehow, but I don't know how. 

I appreciate very much any ideas you have for making the generated `clang` statement output `-g`.

**Comment by martin-frbg** [comment]
Does it work when you add `DEBUG=1` to the `make` commandline ?

**Comment by dugsmith** [comment]
@martin-frbg : thank you for your quick response. No, `DEBUG=1` doesn't change this behavior. As you can see, the `-g` is in the make-generated `clang` statement already, but the one that is actually executed to do the compilation doesn't have `-g` anymore. And it isn't generated with debug symbols either, at least, not that I can see with `nm`.

**Comment by martin-frbg** [comment]
Strange. Is there a reason why you need 0.3.4, or can you try with the current 0.3.10 ?

**Comment by dugsmith** [comment]
Well, I'm stuck with several older dependencies around `OpenCV 3.2.0` in my current build, so I'm trying not to upgrade too far.

I'm actually not sure that the problem is in OpenBLAS itself, but more the internals within Android `clang` that generates the actual `clang` used to compile the file. I was hoping someone here might know what was happening to strip out the `-g` between the original `clang` and the generated `clang`.

**Comment by martin-frbg** [comment]
I have seen some claims that the ndk runs `strip` on the final library by default but keeps an unmodified copy elsewhere in the build tree. (And I notice that while your unexpectedly modified clang line lacks the `-g`, it has several clang options pertaining to
debug, so perhaps the `-g` was only replaced with clang-specific verbiage ?

**Comment by dugsmith** [comment]
@martin-frbg : I think that may be the case. I posted this on the llvm Discord forum, and it seems that when you do `clang -v`, clang shows the "internal" version of the `clang` command it generates from what was submitted to it, along with the `-cc1` mode that does not support `-g`. So the `-g` gets turned into several other arguments that are supposed to add debug symbols. 

I'm still not seeing stack traces in crashes when using `clang -g`, but there may be something else going on. Closing this for now; will update if I learn more.
